File: src/utils/calculations.ts
import type { Habit, HabitLog, Difficulty } from '../types';
import { format, subDays } from 'date-fns';

const DIFFICULTY_MULTIPLIERS: Record<Difficulty, number> = {
  easy: 1.0,
  medium: 1.2,
  hard: 1.5,
  extreme: 2.0,
};

/**
 * Calculate daily completion percentage.
 * Only counts logs that match currently active (non-archived) habits.
 */
export const calculateDailyCompletion = (
  habits: Habit[],
  logs: HabitLog[],
  date: string // YYYY-MM-DD
): number => {
  const activeHabits = habits.filter(h => !h.archived);
  if (activeHabits.length === 0) return 0;

  const activeHabitIds = new Set(activeHabits.map(h => h.id));
  const completedCount = logs.filter(
    l => l.date === date && l.status === 'completed' && activeHabitIds.has(l.habit_id)
  ).length;
  
  return (completedCount / activeHabits.length) * 100;
};

/**
 * Calculate weighted score for a given date.
 * Uses difficulty multipliers. Only counts active habits.
 */
export const calculateWeightedScore = (
  habits: Habit[],
  logs: HabitLog[],
  date: string
): number => {
  const activeHabits = habits.filter(h => !h.archived);
  if (activeHabits.length === 0) return 0;

  const logsForDate = logs.filter(l => l.date === date && l.status === 'completed');
  
  let potentialScore = 0;
  let earnedScore = 0;
  
  activeHabits.forEach(habit => {
    const multiplier = DIFFICULTY_MULTIPLIERS[habit.difficulty] || 1.0;
    potentialScore += multiplier;
    
    const isCompleted = logsForDate.some(l => l.habit_id === habit.id);
    if (isCompleted) {
      earnedScore += multiplier;
    }
  });
  
  if (potentialScore === 0) return 0;
  return (earnedScore / potentialScore) * 100;
};

/**
 * Calculate the Discipline Index (7-day rolling average of weighted scores).
 */
export const calculateDisciplineIndex = (
  habits: Habit[],
  logs: HabitLog[],
  endDate: string = format(new Date(), 'yyyy-MM-dd')
): number => {
  let totalScore = 0;
  const days = 7;
  
  for (let i = 0; i < days; i++) {
    const date = format(subDays(new Date(endDate), i), 'yyyy-MM-dd');
    totalScore += calculateWeightedScore(habits, logs, date);
  }
  
  return Math.round(totalScore / days);
};

/**
 * Calculate current streak for a specific habit (consecutive days completed).
 */
export const getStreak = (
  habitId: string,
  logs: HabitLog[],
  currentDate: string = format(new Date(), 'yyyy-MM-dd')
): number => {
  let streak = 0;
  const habitLogs = logs.filter(l => l.habit_id === habitId && l.status === 'completed');
  
  // Check today
  if (habitLogs.some(l => l.date === currentDate)) {
    streak++;
  }
  
  // Check consecutive days before today
  let checkDate = subDays(new Date(currentDate), 1);
  
  // If today isn't completed, check if yesterday is (streak is alive until EOD)
  if (streak === 0) {
    const yesterdayStr = format(checkDate, 'yyyy-MM-dd');
    if (habitLogs.some(l => l.date === yesterdayStr)) {
      // Don't count today, but start checking from yesterday
    } else {
      return 0; // No streak
    }
  }
  
  while (true) {
    const dateStr = format(checkDate, 'yyyy-MM-dd');
    if (habitLogs.some(l => l.date === dateStr)) {
      streak++;
      checkDate = subDays(checkDate, 1);
    } else {
      break;
    }
  }
  
  return streak;
};

/**
 * Get the actual day of week index (0=Mon, 6=Sun) for a date.
 */
export const getDayOfWeekMondayStart = (date: Date): number => {
  const day = date.getDay(); // 0=Sun, 1=Mon, ...
  return day === 0 ? 6 : day - 1; // Convert to 0=Mon, 6=Sun
};

File: backend/services/calculations.py
"""Port of frontend/src/utils/calculations.ts.

Habits and logs are passed around as plain dicts (matching the JSON shape
Supabase returns) rather than custom classes -- there's no behavior here
that needs a class, just functions over lists of dicts.
"""

from datetime import date, timedelta
from typing import Optional

DIFFICULTY_MULTIPLIERS = {
    "easy": 1.0,
    "medium": 1.2,
    "hard": 1.5,
    "extreme": 2.0,
}


def calculate_daily_completion(habits: list[dict], logs: list[dict], date_str: str) -> float:
    """Percentage of active (non-archived) habits completed on a given day."""
    active_habits = [h for h in habits if not h["archived"]]
    if not active_habits:
        return 0

    active_habit_ids = {h["id"] for h in active_habits}
    completed_count = sum(
        1
        for l in logs
        if l["date"] == date_str and l["status"] == "completed" and l["habit_id"] in active_habit_ids
    )
    return (completed_count / len(active_habits)) * 100


def calculate_weighted_score(habits: list[dict], logs: list[dict], date_str: str) -> float:
    """Difficulty-weighted completion score (0-100) for a given day.
    Harder habits count for more of the day's total than easy ones.
    """
    active_habits = [h for h in habits if not h["archived"]]
    if not active_habits:
        return 0

    completed_habit_ids = {l["habit_id"] for l in logs if l["date"] == date_str and l["status"] == "completed"}

    potential_score = 0.0
    earned_score = 0.0
    for habit in active_habits:
        multiplier = DIFFICULTY_MULTIPLIERS.get(habit["difficulty"], 1.0)
        potential_score += multiplier
        if habit["id"] in completed_habit_ids:
            earned_score += multiplier

    if potential_score == 0:
        return 0
    return (earned_score / potential_score) * 100


def calculate_discipline_index(habits: list[dict], logs: list[dict], end_date_str: Optional[str] = None) -> int:
    """7-day rolling average of daily weighted scores, ending at end_date_str (default today)."""
    end_date = date.fromisoformat(end_date_str) if end_date_str else date.today()
    days = 7
    total_score = 0.0
    for i in range(days):
        d = (end_date - timedelta(days=i)).isoformat()
        total_score += calculate_weighted_score(habits, logs, d)
    return round(total_score / days)


def get_streak(habit_id: str, logs: list[dict], current_date_str: Optional[str] = None) -> int:
    """Current consecutive-day completion streak for one habit.
    A streak survives until end of day: if today isn't done yet but
    yesterday was, the streak is still "alive" and counts from yesterday back.
    """
    current_date = date.fromisoformat(current_date_str) if current_date_str else date.today()
    completed_dates = {l["date"] for l in logs if l["habit_id"] == habit_id and l["status"] == "completed"}

    streak = 0
    if current_date.isoformat() in completed_dates:
        streak += 1

    check_date = current_date - timedelta(days=1)

    if streak == 0:
        yesterday_str = check_date.isoformat()
        if yesterday_str not in completed_dates:
            return 0  # no streak alive

    while check_date.isoformat() in completed_dates:
        streak += 1
        check_date -= timedelta(days=1)

    return streak
